Add basic support & documentation for built-in and plugin-based auth

vdye opened the pull request:
Closes #41

This pull request lays the foundation for specialized auth configuration in the bundle server:

- Commit 1 creates the `AuthResult` type, which the web server will use to relay the output of a custom auth function as an HTTP response
- Commit 2 adds the `--auth-config` type and initial schema; in this commit, it doesn't do much of anything (since there aren't any valid `mode`s yet). Also introduce the `AuthMiddleware` interface, which will be used in both built-in modes and the basis for plugin modes.
- Commit 3 upgrades the version of Go used by the bundle server to 1.20 (necessary for Commit 4)
- Commit 4 adds the `fixed` built-in auth mode, which validates a request's Basic header against a static username/password pair.
- Commits 5 & 6 update the integration tests to include tests of the `fixed` auth mode
- Commit 7 adds the `plugin` auth mode, which allows a user to load (at runtime) a custom implementation of `AuthMiddleware` 

This is a pretty major feature (or the start of one, at least), so I'd really appreciate an extra reviewer focus on making sure the architecture is sound and there aren't any glaring security issues (other than what's already mentioned in the docs, i.e. "don't run untrusted plugins").

derrickstolee commented on cmd/git-bundle-web-server/main.go:
```diff
@@ -29,6 +50,50 @@ func parseAuthConfig(configPath string) (auth.AuthMiddleware, error) {
+		}
+
+		// First, verify plugin checksum matches expected
+		// NEEDSWORK: time-of-check/time-of-use
```
But what else can we do? Can we access the file descriptor of the plugin from `p.Open()` to verify the hash of exactly what we opened? Does the `plugin` library have security checks?

vdye replied:
The `plugin` library has no way of doing that, unfortunately 😞. I considered opening an issue in `golang/go`, but the more I thought about it the less feasible that seemed. [Under the hood](https://github.com/golang/go/blob/134c9b2b171be3515ecef1787ef25435908c17c0/src/plugin/plugin_dlopen.go#L9-L33), Go is using `dlopen` & `dlsym` to load the plugin symbols, so even if it _did_ verify the checksum before running the init tasks of the Go package, someone could still (carefully) overwrite the file contents before invoking the `initializer` symbol. 

If we _really_ wanted to do this, we'd probably need to write our own version of `plugin` that:

1. opens the file with `dlopen()`
2. locks it with `flock()`
3. checks the checksum
4. loads & runs the init tasks
5. loads the requested symbol with `dlsym()`
6. invokes the symbol 
7. releases the lock with `flock()`
8. closes the file

Nearly all of that would need to be done with C bindings and would need platform-specific implementations for Mac & Linux due to slight differences in `flock()` (and would further distance us from Windows support, if that matters?). To be honest, that doesn't necessarily seem worth it. 

So my next question is, should we remove/change the corresponding warning in `docs/technical/auth-config.md` to drop the specific mention of TOCTOU and just recommend general safety around custom plugins?

vdye replied:
If no one's opposed, I'll leave the TOCTOU mention - seems better to inform users to watch for/mitigate it than to pretend it's not an issue.

derrickstolee commented on docs/technical/auth-config.md:
```diff
@@ -25,11 +25,48 @@ The JSON file contains the following fields:
         </tr>
         <tr>
-            <td><code>parameters</code></td>
+            <td><code>path</code> (<code>plugin</code> mode only)</td>
```
Interesting that these are not part of `parameters` for `plugin` mode. Could we make that swap? It can have a `pluginParameters` value inside if more customization is needed.

derrickstolee replied:
My initial thought was "we have a few named parameters that apply to every mode, and anything else fits in `parameters`." If you anticipate that `basic` is the outlier and really everyone will be using a plugin, then having these values where they are makes more sense.

derrickstolee reviewed (commented):
Neat plugin interface. I had some comments that might help as you harden things.

vdye commented on cmd/git-bundle-web-server/main.go:
```diff
@@ -28,9 +56,21 @@ func main() {
+			if err != nil {
+				logger.Fatalf(ctx, "Invalid auth config: %w", err)
+			}
+			if middleware == nil {
+				panic("Middleware is nil, but no error was returned")
+			}
```
If this condition is reached, the auth config is set (and isn't empty) _and_ the initializer was successfully called _and_ it didn't return an error but still returned `nil` for the middleware. This is meant more as a sanity check than anything else: everything about the user's setup implies they want to use auth middleware, but the value is nil; that, to me, implies there's an issue with how the initializer is written. 

If a user is expecting the auth to be loaded but it's `nil` (without another error), that feels like it warrants an error. I can change it to to a `log.Fatalf()` if you think that would be better? I'll also add a comment explaining it better.

derrickstolee replied:
I think of `panic()` as the equivalent of `BUG()`, not `die()`. That's probably wrong, though, since there really isn't another way to exist a Go process quickly. Feel free to leave this as-is.

vdye replied:
I ended up switching to `log.Fatalf()`, since I have generally been using `panic()` as an analogue for `BUG()`. Also added the comment!

vdye commented on cmd/git-bundle-web-server/main.go:
```diff
@@ -27,8 +27,10 @@ func parseAuthConfig(configPath string) (auth.AuthMiddleware, error) {
+		return auth.NewBasicAuth(config.Parameters)
 	default:
-		return nil, fmt.Errorf("unrecognized auth mode '%s'", config.AuthMode)
+		return nil, fmt.Errorf("unrecognized auth type '%s'", config.AuthMode)
```
Ah this is a holdover from an older iteration, it should remain `mode`.

vdye commented on cmd/git-bundle-web-server/bundle-server.go:
```diff
@@ -107,6 +113,13 @@ func (b *bundleWebServer) serve(w http.ResponseWriter, r *http.Request) {
 
+	if b.authorize != nil {
+		authResult := b.authorize(r, owner, repo)
+		if authResult.WriteResponse(w) {
+			return
+		}
```
I originally had it as the latter, but I wasn't thrilled with needing to re-check `Authorized()` in `WriteResponse()` (because if I was doing that anyway, why even have the original `Authorized()` check?). 

I'll add a comment, but I'll also try to come up with a better name (`AuthorizeOrWriteResponse()`? or maybe something simpler like `ApplyResult()`).

mjcheetham replied:
If we wanted to give auth middleware full control of the pipeline then really we should really only expect a boolean "return bundles or exit?" and have the middleware handle the entire response?

```go
if b.authorize != nil {
    if !b.authorize(r, w, owner, repo) {
        return
    }
}
```

I assume the only expected front-end/consumer of the bundle server endpoints is Git (never a browser etc)?

vdye replied:
It took a few iterations, but I've made the `AuthResult`s a bit more generic (where `Allow()` == "apply headers & serve content" and `Deny()` == "apply headers & respond with given 4XX code"). I'd like to keep the middleware at least a bit sandboxed (it helps keep the bundle server aligned with what Git expects/supports and IMO makes plugin writing simpler), but if there's anything else that might be valuable to add to the current interface, I'm happy to update!

vdye commented on docs/technical/auth-config.md:
```diff
@@ -25,11 +25,48 @@ The JSON file contains the following fields:
+            </td>
+        </tr>
+        <tr>
+            <td><code>initializer</code> (<code>plugin</code> mode only)</td>
```
This should be `sha256` not `initializer`.

mjcheetham reviewed (commented):
One overarching question I had is do we want to relax the plugin interface _even more_? From the bundle server's perspective the only real decision it needs to make is "should I return bundles or not".

Splitting out unauthorized 401, forbidden 403(404?) and success (200) means plugins can only use those response options.

vdye commented on pkg/auth/auth-result.go:
```diff
@@ -0,0 +1,71 @@
+		// Do not write a response, we're authorized
+		return false
+	case unauthorized:
+		for _, h := range a.headers {
+			w.Header().Add(h.key, h.value)
+		}
```
The `Header()` structure is actually a `map` with `string` keys and `[]string` values. I'm not 100% sure, but I think (based on [this](https://github.com/golang/go/blob/945a2b17f3b47b4e0bda9a2a92186412b5704c9a/src/net/http/header.go#L173-L187)) the headers will be first sorted alphabetically, then in order of addition (for multiples of the same header).

mjcheetham replied:
> I'm not 100% sure, but I think (based on [this](https://github.com/golang/go/blob/945a2b17f3b47b4e0bda9a2a92186412b5704c9a/src/net/http/header.go#L173-L187)) the headers will be first sorted alphabetically, then in order of addition (for multiples of the same header).

The order of (repeated) auth headers matters.

```text
WWW-Authenticate: Bearer authorize="idp.example.com/oauth"
WWW-Authenticate: Basic realm="example.com"
```

..may be treated differently by clients compared to..

```text
WWW-Authenticate: Basic realm="example.com"
WWW-Authenticate: Bearer authorize="idp.example.com/oauth"
```

Of course if order of repeated headers matters, the middleware could always embed multiple header values in the same field:

```text
WWW-Authenticate: Bearer authorize="idp.example.com/oauth", Basic realm="example.com"
```

If it's possible, maintaining the order the middleware gave would probably be the least surprising behaviour.

vdye replied:
Repeated headers will be kept in the order that they're provided by the middleware (they're added to `Header()` in order [here](https://github.com/git-ecosystem/git-bundle-server/blob/40a02483218e97df418ce868f4230e2722173dc4/pkg/auth/auth-result.go#L36-L38), then added to the response in order [here](https://github.com/golang/go/blob/945a2b17f3b47b4e0bda9a2a92186412b5704c9a/src/net/http/header.go#L211-L223).

Since the `AuthResult` methods will be in a public package, I'll also add formal function documentation to make that clear.

mjcheetham replied:
> Repeated headers will be kept in the order that they're provided by the middleware

Ah, then that should be fine!

vdye commented on docs/technical/auth-config.md:
````diff
@@ -38,3 +38,69 @@ The JSON file contains the following fields:
+```json
+{
+    "name": "basic",
+    "parameters": {
+        "username": "admin",
+        "password": ""
+    }
````
For this particular middleware, yes. Maybe I should rename this mode to `singleuser` to reflect that? `basic` is more of a description of the protocol, and doesn't reflect the user management aspect of the middleware.

vdye replied:
I ended up on `fixed` as the new name, but I'm happy to change it to something else if anyone has a better suggestion. 🙂

mjcheetham replied:
#bikeshedding... `fixed` is fine with me

vdye commented on pkg/auth/auth-result.go:
```diff
@@ -0,0 +1,71 @@
+		}
+		w.WriteHeader(http.StatusUnauthorized)
+		return true
+	case forbidden:
+		// Return a 404 status even though the issue is that the user is
+		// forbidden so we don't indirectly reveal which repositories are
+		// configured in the bundle server.
+		w.WriteHeader(http.StatusNotFound)
+		return true
```
The middleware returns one of the valid response types (currently `Authorized` -> 200, `Unauthorized` (which can have configured headers) -> 401, and `Forbidden` -> 404). So, the middleware chooses the response, but the bundle server decides what to do with that.

Generally speaking, I took a _very_ restrictive approach with this design in an attempt to keep users from writing a plugin that unintentionally reveals too much information; that might be more frustrating than helpful, though. I could instead allow middleware to specify _any_ 4XX code (with optional headers). Then, there's basically two `AuthResult` states: 

1. User is authorized, serve the content
2. User is not authorized, send the specified 4XX error & headers

Does that sound reasonable, or do you think it should be expanded further?

mjcheetham replied:
> * User is authorized, serve the content
> * User is not authorized, send the specified 4XX error & headers

These seem reasonable, yes.

A middleware author may reasonably want to 200, 401, 403 or 404 with headers for auth.

By the way, technical auth middleware may also want to add headers in the case of a 200 also, for example cookies, DPoP nonces, or [`Authentication-Info`](https://datatracker.ietf.org/doc/html/rfc7615#section-3).

vdye replied:
Won't Git drop headers it receives from a 200 response? That's the main reason I didn't include a `headers` arg in `Accept()`.

mjcheetham replied:
> Won't Git drop headers it receives from a 200 response?

Today, yes 😉 

I highly suspect a future where 200s need headers too (passed to credential helpers as part of `store` requests). When Microsoft first tried to push us to access token binding it would require reading headers on 200 and using that data to mint tokens in subsequent requests.

DPoP (draft spec, admittedly) requires headers on 200:
https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop#section-8.2

vdye replied:
Added `headers` to (the once-again-renamed) `Allow()`!

vdye commented on pkg/auth/middleware.go:
```diff
@@ -14,3 +19,57 @@ import (
 	Authorize(r *http.Request, owner string, repo string) *AuthResult
 }
+
+/* Built-in auth modes */
+// Authorize users with a credentials matching a static username/password pair
+// that applies to the whole server.
+type basicAuth struct {
+	usernameHash [32]byte
+	passwordHash [32]byte
+}
+
+type basicAuthParams struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
```
Good point. I could add a script to the repo that generates the requested (built in) auth config based on interactive user prompts, then install that script with the bundle server. Or maybe a `git-bundle-server` command? 

I'll update the schema to require users to specify the hash, then also open an issue to add that script/command.

vdye replied:
Updated the `fixed` mode & opened #52

vdye commented on docs/technical/auth-config.md:
```diff
@@ -0,0 +1,40 @@
+            </td>
+        </tr>
+    </tbody>
+</table>
```
This plaintext format would work well in the manpages, where it's probably valuable to have auth config schema information anyway. I'll add this there.

vdye replied:
Updated the `git-bundle-web-server` man page with an `AUTH CONFIG` section containing this info + some examples. `docs/technical/auth-config.md` is still more detailed, but the manpage is still fairly informative.

vdye commented:
@derrickstolee @mjcheetham I _think_ I've addressed all of the feedback so far. The PR has changed a fair amount since the first round of reviews (added integration tests, manpage docs, and better in-code documentation for `pkg/auth`; also, changed `AuthResult` pretty substantially), so please let me know if you have any questions or otherwise think something should be changed! 🙇‍♀️

ldennington commented on internal/auth/middleware_test.go:
```diff
@@ -0,0 +1,164 @@
+	expectedResponseCode    int
+	expectedHeaders         http.Header
+}{
+	{
```
Really thorough set of test cases here!

ldennington reviewed (approved):
Looks really good! Very impressed all-up, but the plugin work is especially awesome.

mjcheetham commented on internal/auth/middleware.go:
```diff
@@ -0,0 +1,78 @@
+
+	// Check for invalid username characters
+	if strings.Contains(params.Username, ":") {
+		return nil, fmt.Errorf("username contains a colon (\":\")")
```
I'd imagine to avoid issues encoding this in a URL?

```
protocol://user:pass@host/path
```

Same with `@`; you'd need to escape it.

vdye replied:
It's part of [RFC 7617](https://www.rfc-editor.org/rfc/rfc7617), which defines Basic auth:

> Furthermore, a user-id containing a colon character is invalid, as the first colon in a user-pass string separates user-id and password from one another; text after the first colon is part of the password. User-ids containing colons cannot be encoded in user-pass strings.
